File: leetcode/easy/best_time_to_bns_stock_ii.go
func maxProfit(prices []int) int {
    isUp := false
    
    total := 0
    currentLowest := 0
    
    prev := prices[0]
    
    for i := 1; i < len(prices); i++ {
        if prices[i] > prev {
            if (!isUp) {
                isUp = true
                currentLowest = prev
            }
        } else if(isUp) {
            isUp = false
            total += prev - currentLowest
        }
        
        prev = prices[i]
    }
    
    if (isUp) {
        total += (prev - currentLowest)
    }
    
    return total
}

File: leetcode/best_time_to_bns_stock_ii.rb
# @param {Integer[]} prices
# @return {Integer}
def max_profit(prices)
    is_going_up = false
    
    total = 0
    current_lowest = 0
    
    prev = prices[0]
    
    prices.each do |price|
        if is_going_up
            if price <= prev
                is_going_up = false
                total += prev - current_lowest
            end
        elsif price > prev
            is_going_up = true
            current_lowest = prev
        end
        
        prev = price
    end
    
    if is_going_up
        total += prev - current_lowest
    end
    
    return total
end